### PR TITLE
Integration work to support compiling R2R images using crossgen2

### PR DESCRIPTION
**Pull request by fadimounir**
These changes enable the SDK to download and use the crossgen2 package and use crossgen2 as the ReadyToRun compiler. Unlike crossgen, crossgen2 is a tool in its own package, and no longer part of the runtime package. 

The supported scenarios for crossgen2 are win-x64->win-x64 and linux-x64->linux-x64 compilations, and is only enabled if the PublishReadyToRunUseCrossgen2 property is set to true.

Also, crossgen2 currently still does not have the ability to emit native symbols for the compiled ReadyToRun images, so for those we will use crossgen from the runtime pack.

### CODE REVIEW

**Comment by fadimounir** [comment]
cc @dotnet/crossgen-contrib

**Comment by fadimounir** [comment]
@nguerrera, this is still work in progress, but could you please take a look and give me some feedback?

File: src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs [hunk 1]
```diff
@@ -400,28 +442,47 @@ private bool GetCrossgenComponentsPaths()
             {
                 if (_targetArchitecture == Architecture.Arm)
                 {
-                    _crossgenPath = Path.Combine(_packagePath, "tools", "x86_arm", "crossgen.exe");
-                    _clrjitPath = Path.Combine(_packagePath, "runtimes", "x86_arm", "native", "clrjit.dll");
-                    _diasymreaderPath = Path.Combine(_packagePath, "runtimes", _runtimeIdentifier, "native", "Microsoft.DiaSymReader.Native.x86.dll");
+                    if (RuntimeInformation.OSArchitecture == _targetArchitecture)
```
Comment (fadimounir) [inline src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs]: This is a fix to a separate bug... maybe it should go into a separate PR? Let me know

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets [hunk 2]
```diff
@@ -21,6 +21,16 @@ Copyright (c) .NET Foundation. All rights reserved.
   <UsingTask TaskName="ProcessFrameworkReferences" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
   <UsingTask TaskName="ResolveAppHosts" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
 
+  <ItemGroup>
+    <!-- TODO: This should be moved to Microsoft.NETCoreSdk.BundledVersions.props (src/redist/targets/GenerateBundledVersions.targets in dotnet/core-sdk) -->
+    <KnownCrossgen2Pack Include="Microsoft.NETCore.App.Crossgen2"
```
Comment (fadimounir) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets]: https://github.com/dotnet/core-sdk/pull/6227

File: src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs [hunk 3]
```diff
@@ -178,7 +219,8 @@ private void ProcessInputFileList(ITaskItem[] inputFiles, List<ITaskItem> imageC
                 // unless an explicit PublishReadyToRunEmitSymbols flag is enabled by the app developer. There is also another way to profile that the runtime supports, which does
                 // not rely on the native PDBs/Map files, so creating them is really an opt-in option, typically used by advanced users.
                 // For debugging, only the IL PDBs are required.
-                if (EmitSymbols)
+                // TODO: Need to figure out if we'll add the native symbol emission feature to crossgen2, or use crossgen to generate them
```
Comment (fadimounir) [inline src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs]: There's a conversation going on around the native symbols feature. Most likely, it's looking like we'll use crossgen for it, as this won't be implemented in crossgen2 for the .NET 5.0 timeframe

File: src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs (hunk 1 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs]: What's the bug? Yes, separate PR. Could be helpful if we need to port to other branches?

Comment (fadimounir) [reply src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs]: The bug is that we couldn't run the native arm/arm64 crossgen and publish from a Windows arm/arm64 device. I'll move that to a separate PR.

**Comment by fadimounir** [comment]
@nguerrera Could you please take a look? This PR is finally ready, and I expect the crossgen2 CI tests to pass after the latest version update I put in global.json

File: src/Tasks/Common/Resources/Strings.resx [hunk 4]
```diff
@@ -604,4 +604,8 @@ The following are names of parameters or literal values and should not be transl
     <value>NETSDK1125: Publishing to a single-file is only supported for netcoreapp target.</value>
     <comment>{StrBegin="NETSDK1125: "}</comment>
   </data>
+  <data name="Crossgen2RequiresSelfContained" xml:space="preserve">
+    <value>NETSDK1126: Publishing ReadyToRun using Crossgen2 is only supported for self-contained applications for now.</value>
```
Comment (nguerrera) [inline src/Tasks/Common/Resources/Strings.resx]: I would remove the "for now".

File: src/Tasks/Common/Resources/Strings.resx [hunk 5]
```diff
@@ -604,4 +604,8 @@ The following are names of parameters or literal values and should not be transl
...
+    <comment>{StrBegin="NETSDK9999: "}</comment>
```
Comment (nguerrera) [inline src/Tasks/Common/Resources/Strings.resx]: This has the wrong version, I though I added validation for this. Is there a test failing on it.

File: src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs [hunk 6]
```diff
@@ -15,31 +12,43 @@ namespace Microsoft.NET.Build.Tasks
...
 
         private bool IsPdbCompilation => !String.IsNullOrEmpty(_createPDBCommand);
 
-        protected override string ToolName => _crossgenPath;
+        protected override string ToolName
+        {
+            get
+            {
+                // NOTE: Crossgen2 does not yet support emitting native symbols. We use crossgen instead for now.
+                if (UseCrossgen2)
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs]: .nit: we put braces around all if/else/loop blocks

File: src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs [hunk 7]
```diff
@@ -48,23 +57,39 @@ public RunReadyToRunCompiler()
...
         {
-            _crossgenPath = CrossgenTool.ItemSpec;
-            _clrjitPath = CrossgenTool.GetMetadata("JitPath");
-            _diasymreaderPath = CrossgenTool.GetMetadata("DiaSymReader");
+            _createPDBCommand = CompilationEntry.GetMetadata("CreatePDBCommand");
 
-            if (!File.Exists(_crossgenPath) || !File.Exists(_clrjitPath))
+            if (CrossgenTool == null && Crossgen2Tool == null)
             {
                 return false;
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs]: Already like this, so not high pri, but what is the error when this returns false. Should we consider logging the individual cases? I suppose these are mainly thing that are not expected to actually happen?

File: src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs [hunk 8]
```diff
@@ -99,13 +124,25 @@ private string GetAssemblyReferencesCommands()
...
+                    result.AppendLine($"-r \"{reference}\"");
             }
 
             return result.ToString();
         }
 
         protected override string GenerateResponseFileCommands()
+        {
+            // NOTE: Crossgen2 does not yet support emitting native symbols. We use crossgen instead for now.
+            if (IsPdbCompilation)
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs]: nit: braces

File: src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs [hunk 9]
```diff
@@ -135,6 +172,23 @@ protected override string GenerateResponseFileCommands()
...
 
+        private string GenerateCrossgen2ResponseFile()
+        {
+            StringBuilder result = new StringBuilder();
+
+            result.AppendLine("-O");
+            result.AppendLine($"--jitpath:\"{Crossgen2Tool.GetMetadata("JitPath")}\"");
+            result.Append(GetAssemblyReferencesCommands());
+            result.AppendLine($"--out:\"{_outputR2RImage}\"");
+            if (!String.IsNullOrEmpty(Crossgen2ExtraCommandLineArgs))
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs]: nit: braces

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs [hunk 10]
```diff
@@ -253,27 +184,69 @@ public void It_warns_when_targetting_netcoreapp_2_x()
...
+            TestProjectPublishing_Internal("LibraryProject2", targetFramework, isSelfContained:true, makeExeProject: false);
+        }
+
+        [Theory]
+        [InlineData("netcoreapp5.0")]
+        public void It_can_publish_readytorun_using_crossgen2(string targetFramework)
+        {
+            // Crossgen2 only supported for Linux/Windows x64 scenarios for now
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX) || RuntimeInformation.OSArchitecture != Architecture.X64)
+                return;
```
Comment (nguerrera) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs]: nit: braces

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs [hunk 11]
```diff
@@ -253,27 +184,69 @@ public void It_warns_when_targetting_netcoreapp_2_x()
...
         }
 
-        private void TestLibProjectPublishing_Internal(string projectName, bool isSelfContained, string targetFramework)
+        [Theory]
+        [InlineData("netcoreapp5.0")]
+        public void It_only_supports_selfcontained_when_using_crossgen2(string targetFramework)
         {
+            // Crossgen2 only supported for Linux/Windows x64 scenarios for now
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX) || RuntimeInformation.OSArchitecture != Architecture.X64)
+                return;
```
Comment (nguerrera) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs]: nit: braces

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs [hunk 12]
```diff
@@ -253,27 +184,69 @@ public void It_warns_when_targetting_netcoreapp_2_x()
...
+            var testProjectInstance = _testAssetsManager.CreateTestProject(testProject);
+
+            var publishCommand = new PublishCommand(Log, Path.Combine(testProjectInstance.Path, testProject.Name));
+            publishCommand.Execute()
+                .Should()
+                .Fail()
+                .And.HaveStdOutContainingIgnoreCase("NETSDK1126");
+        }
+
+        #region Helpers
```
Comment (nguerrera) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs]: nit: avoid regions.

File: src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs [hunk 13]
```diff
@@ -17,14 +17,24 @@ namespace Microsoft.NET.Build.Tasks
 {
     public class PrepareForReadyToRunCompilation : TaskBase
     {
+        internal struct CrossgenToolInfo
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs]: Looks like all uses of this are private, so this can be private. 

nit: consider moving this lower, away from the public task parameters.

File: src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs [hunk 14]
```diff
@@ -79,40 +93,105 @@ protected override void ExecuteCore()
...
                 Log.LogError(Strings.ReadyToRunNoValidRuntimePackageError);
                 return;
             }
 
-            if (!ExtractTargetPlatformAndArchitecture(_runtimeIdentifier, out string targetPlatform, out _targetArchitecture) ||
-                !ExtractTargetPlatformAndArchitecture(hostRuntimeIdentifier, out string hostPlatform, out Architecture hostArchitecture) ||
+            if (ReadyToRunUseCrossgen2)
+            {
+                if (!ValidateCrossgen2Support())
+                    return;
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs]: nit: braces

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs]: ok, I give up on commenting on all of them. :)

File: src/Tasks/Common/Resources/Strings.resx (hunk 5 above)
Comment (nguerrera) [reply src/Tasks/Common/Resources/Strings.resx]: Yes, there is a test failing with "Expected string to start with \n"{StrBegin=\"NETSDK1126: \"}" because localization instructions should indicate invariant error code as preceding translatable message., but \n"{StrBegin=\"NETSDK9999: \"}" differs near "999" (index 17)."

**Comment by nguerrera** [comment]
The more interesting test failures seem to be something trying to use a RID as a target framework. Trying to find where that is.

**Comment by nguerrera** [comment]
Looks like this is the cause:

``` diff
diff --git a/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs b/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs
index 017951c7c..1c92ec8f2 100644
--- a/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs
+++ b/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishReadyToRun.cs
@@ -27,7 +27,7 @@ public void It_only_runs_readytorun_compiler_when_switch_is_enabled(string targe
             var projectName = "CrossgenTest1";

             var testProject = CreateTestProjectForR2RTesting(
-                EnvironmentInfo.GetCompatibleRid(targetFramework),
+                targetFramework,
                 projectName,
                 "ClassLib");

@@ -54,7 +54,7 @@ public void It_creates_readytorun_images_for_all_assemblies_except_excluded_ones
             var projectName = "CrossgenTest2";

             var testProject = CreateTestProjectForR2RTesting(
-                EnvironmentInfo.GetCompatibleRid(targetFramework),
+                targetFramework,
                 projectName,
                 "ClassLib");
```

File: src/Tasks/Common/Resources/Strings.resx (hunk 4 above)
Comment (nguerrera) [reply src/Tasks/Common/Resources/Strings.resx]: You never know how plans will change.

**Comment by nguerrera** [comment]
Can you update the top-level description of the PR with more information about what this does, how to use it, etc.?

File: src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs (hunk 14 above)
Comment (fadimounir) [reply src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs]: Added braces to a bunch of places

File: src/Tasks/Common/Resources/Strings.resx (hunk 5 above)
Comment (fadimounir) [reply src/Tasks/Common/Resources/Strings.resx]: Good catch. Yes this was an oversight from a previous attempt to use 9999 as a code. Fixed

File: src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs (hunk 7 above)
Comment (fadimounir) [reply src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs]: Correct, this is not expected to actually happen. If we fail to resolve the crossgen tool earlier in the pipeline (PrepareForReadyToRunCompilation), the task will show meaningful messages.

**Review by nguerrera — APPROVED** [review]

**Comment by nguerrera** [comment]
master is now preview 2. I don't know if there's any process yet for 5.0 preview 1. Does this need to be in preview 1?

**Comment by fadimounir** [comment]
It can be in preview 2
